hibernate-core/src/main/java/org/hibernate/persister/entity/JoinedSubclassEntityPersister.java

```diff
@@ -1240,7 +1240,7 @@ public FilterAliasGenerator getFilterAliasGenerator(String rootAlias) {
 
 	@Override
 	public TableDetails getMappedTableDetails() {
-		return getTableMapping( getTableMappings().length - 1 );
+		return getTableMapping( getTableMappings().length - ( tableSpan - coreTableSpan ) - 1 );
```

Comment:
The last entry is _usually_ the "current" table, i.e. the table of the subtype represented by the current persister, **unsless there are secondary tables**: than the list is like `["root", "child1", "subchild1", "secondary_table"]`. That is why we need to subtract the number of secondary tables `tableSpan - coreTableSpan` first.

Also, `coreTableSpan - 1` doesn't work because we allow subclasses having the same table name with an explicit `@DiscriminatorColumn`, but when building table mappings we merge duplicates based on name (e.g. `coreTableSpan` is 2 but `tableMappings.length` is 1).